mchf-eclipse/drivers/audio/cw/cw_gen.c

```diff
@@ -579,7 +579,10 @@ static void CwGen_CheckKeyerState(void)
 
 	if (!ts.cw_text_entry && ps.cw_state == CW_IDLE)
 	{
-		if (ts.txrx_mode == TRX_MODE_TX && DigiModes_TxBufferHasData() && !ps.sending_char &&
+	    // FIXME: I am not sure how this is supposed to work, but without check for DEMOD_CW it happily eats
+	    // all keyboard input for modes such as RTTY or PSK...
+	    // it seems to be intended for transmitting morse via keyboard input...
+		if (ts.dmod_mode == DEMOD_CW && ts.txrx_mode == TRX_MODE_TX && DigiModes_TxBufferHasData() && !ps.sending_char &&
 				(! (ps.port_state & CW_END_PROC) && ps.space_timer < ps.space_time - ps.dah_time))
```

Comment:
Hi Danilo, I handled already it, and it would be nice if you checkout this file back. It would be much easy for me to rebase later. Thank you...

Reply:
If, it's a problem don't worry I will do it, not a big deal

Reply:
> What did you handle? 

I implemented kind of multiplexer for digi_buffer, so only consumer which was set can remove from buffer, also only digi_buffer send character to UI_buffer on input or on outputing. It gives chance to broke infinite loop when CW produces char -> put it into buffer -> and remove the same input from buffer.